### DISCUSSION TITLE
Support for dynamic fields equivalent to patternProperties in JSON Schema

### DISCUSSION THREAD

**Discussion by LSerranoPEReN** [pydantic/pydantic#3303] [Feature Request]
Our team uses Pydantic through FastAPI to check and process user JSON inputs.
In our data model, some keys are static but other are dynamic keys, that follows a pattern, provided by the user.

With JSON Schema validating such inputs can be done with [patternProperties](https://json-schema.org/understanding-json-schema/reference/object.html#pattern-properties).
As far as I understand support for dynamic fields in Pydantic is very limited: https://github.com/samuelcolvin/pydantic/issues/1809

It will be great to have a support of dynamically named fields in Pydantic.

---- 

I have a small working implementation https://github.com/LSerranoPEReN/pydantic/commit/abfc16c5084d08465dc800becb093c99df403ab8 that can be converted to a PR if such feature is approved.

Here is an usage example which work our current implementation. It uses a new function named `PatternField` to declare such fields.

### Declaration
```python3
DistroFamily = ForwardRef('DistroFamily')
class DistroFamily(BaseModel):
    creation_date: Optional[int] = Field(alias="_creation_date")
    children: Optional[DistroFamily] = PatternField("^[^_]")
DistroFamily.update_forward_refs()
```

### Schema generation
`print(DistroFamily.schema_json(indent=4))`:
```json
{
    "$ref": "#/definitions/DistroFamily",
    "definitions": {
        "DistroFamily": {
            "title": "DistroFamily",
            "type": "object",
            "properties": {
                "_creation_date": {
                    "title": " Creation Date",
                    "type": "integer"
                }
            },
            "patternProperties": {
                "^[^_]": {
                    "$ref": "#/definitions/DistroFamily"
                }
            }
        }
    }
}
```

### Data processing
```python3
debian_data = {
    '_creation_date': 1993,
    'ubuntu' : {
        '_creation_date': 2004,
        'kubuntu': {},
        'chromium_os': {
            '_creation_date': 2009
        }
    }
}

print(DistroFamily(**debian_data).children)
>>> {'ubuntu': DistroFamily(creation_date=2004, children={'kubuntu': DistroFamily(creation_date=None, children={}), 'chromium_os': DistroFamily(creation_date=2009, children={})})}
```

**Comment by certino-simonchapman** [comment]
did anything come of this, because this is so needed. Have a usecase right now, so really hoping someone took this seriously.

**Comment by samuelcolvin** [comment]
There's no built in support for pattern properties in v1 and it's too late to add it now.

Happy to consider adding it it in v2, it would require a specific validator in rust.

See https://github.com/pydantic/pydantic-core/issues/305

**Reply by certino-simonchapman** [reply]
that would be great thanks @samuelcolvin - would be a great feature.

**Comment by rafalkrupinski** [comment]
I was able to do it  directly in the model (pydantic v2).

It follows `model_extra` pattern, you add a synthetic field, with an extra annotation (`PropertyPattern`) and a 'before' model validator that reshapes the input data.

The validator checks all fields for that annotation, then moves all matching properties from the root dict to that field.

I've done it as a base class. The example class models `paths` object from OpenAPI.

```python
class PropertyPattern:
    def __init__(self, pattern: str) -> None:
        self.pattern = pattern

class ModelWithPatternProperties(BaseModel):
    @model_validator(mode='before')
    @classmethod
    def validate(cls, value: typing.Any, info: ValidationInfo):
        for field_name, field_info in cls.model_fields.items():
            pattern_anno = find_annotation_optional(field_info.metadata, PropertyPattern)
            if not pattern_anno:
                continue

            pattern = re.compile(pattern_anno.pattern)

            pattern_props = {}
            for key, item in value.items():
                if pattern.search(key):
                    pattern_props[key] = item
            for key in pattern_props:
                del value[key]
            value[field_name] = pattern_props

        return value

class Paths(ModelWithPatternProperties):
    paths: Annotated[dict[str, PathItem], Field(default_factory=dict), PropertyPattern('^/')]
```
